### DISCUSSION TITLE
Help with Lua

### DISCUSSION THREAD

**Discussion by thaihugo** [jcorporation/mympd-scripts#11] [Q&A] [answered]
myMPD version: 8.1.5
MPD: 0.22.4
Debian buster on Raspberry pi 4

Hi,

I need some help on using Lua. The following scripts where working as intended when the interpreter appeared in previous versions, but now they have some issues I would like to understand.

Both this script run in alternative way. One replace the other at different time of the day.

`-- {"order":1,"arguments":[]}

mympd_api("MYMPD_API_PLAYER_OPTIONS_SET", "jukeboxMode", "1")
mympd_api("MYMPD_API_PLAYER_OPTIONS_SET", "jukeboxPlaylist", "Database")
mympd_api("MYMPD_API_PLAYER_OPTIONS_SET", "jukeboxUniqueTag", "Album")
mympd_api("MYMPD_API_PLAYER_OPTIONS_SET", "jukeboxLastPlayed", 240)
--mympd_api("MYMPD_API_VOLUME_SET",  40)
mympd_api("MYMPD_API_QUEUE_CLEAR")
mympd_api("MYMPD_API_PLAYER_STOP")
`
This script has the following problems:

- the player doesn't stop,
- Uniqueness is set to song instead of album.

`-- {"order":2,"arguments":[]}

--mympd = { _version = "0.2.0" }

mympd_api("MYMPD_API_PLAYER_STOP")
mympd_api("MYMPD_API_PLAYER_OPTIONS_SET", "jukeboxMode", "1")
mympd_api("MYMPD_API_PLAYER_OPTIONS_SET", "jukeboxPlaylist", "Nanas awesome mix")
mympd_api("MYMPD_API_PLAYER_OPTIONS_SET", "jukeboxUniqueTag", "Album")
mympd_api("MYMPD_API_PLAYER_OPTIONS_SET", "jukeboxLastPlayed", 48)
--mympd_api("MYMPD_API_PLAYER_OPTIONS_SET", "jukebox_unique_tag", "Album")
mympd_api("MYMPD_API_QUEUE_CLEAR")
mympd_api("MYMPD_API_QUEUE_SHUFFLE")
mympd_api("MYMPD_API_PLAYER_STOP")
`

This script has the following problems:

- uniqueness is set to song instead of album,
- Without the shuffle, the same playlist will be issued every day. This is not the case with the other script. 

Also, this one does stop at the last instruction.

Could you please point me in the right direction ? I don't understand why this is working in some cases and not in others. Maybe there is indentation rules in lua?

thanks for any help

Thaihugo

**Comment by jcorporation** [comment]
The scripts are looking good. You can combine the MYMPD_API_PLAYER_OPTIONS_SET calls in one call, just append the key/value pairs.
`mympd_api("MYMPD_API_PLAYER_OPTIONS_SET", "jukeboxMode", "1", "jukeboxPlaylist", "Database", "jukeboxUniqueTag", "Album", "jukeboxLastPlayed", 240)`

> Uniqueness is set to song instead of album.

This is a bug - the ui shows always song, even if you change it in the ui. This is now fixed with commit 39cbdb17fba21a0cf000848297e25b08d02ad0c4.

Please provide debug logs of myMPD and verbose logs of MPD.

> Without the shuffle, the same playlist will be issued every day. This is not the case with the other script.

I do not understand, can you describe this more verbose? How many songs the jukebox adds? I do not understand why do you need the queue shuffle?

**Comment by thaihugo** [comment]
Thanks. My understanding is that:

- The uniqueness parameter is set but was not displayed correctly and it is now corrected in a commit.
- You want the debug logs for the fact the played doesn't stop even after the player_stop api call. I'll provide them for this action.
- You need more info for the randomness issue:
 
 The playlist used here is small. Probably 20-30 albums. When using this exact script without the queue shuffle, the same songs will appear in the list, everyday. With the other script working on the full library, I will get new songs each time it is launched.
If I input the jukebox parameters in the interface manually, I get normal randomness. The script suppress that randomness.

I always had randomness problems, like a few song seems to be repeated too much statistically compared to my music collection. But the small playlist is definitely not randomized when this script insert it in the jukebox.

**Comment by thaihugo** [comment]
Here are the log. 


The issue with the player not stopping is happening in this log. Also, the second playlist has a new selection which is not random 

actions:

- Run mympd which was under the second script configuration.
- Launch the first script (hugo.lua). Music plays instead of stopping.
- Launch the second script (nana.lua). Get the same list for the same playlist (changes in the playlist gives a different result). The clear_queue api call seems ignored too.

hope this help you see the issue.

thanks

Thaihugo

**Comment by jcorporation** [comment]
Thanks for the logs. I will look into it.

Concerning the randomness issue:
- 20-30 albums in the playlist
- uniq tag is album
- you add 20 songs to the queue

This could not produce a random list. The jukebox uses reservoir sampling to create the internal jukebox queue with 50 entries. From this internal queue the mpd queue is filled. Because you have uniq tag set to album, from each album only the first song is considered to be added and the first albums are always added because the jukebox tries to fill up to 50 entries. 

You should not use uniq tag album if there are only a few albums in the playlist. The log shows the respectively messages:
```
DEBUG    mympdapi  /tmp/myMPD-8.1.5/src/mpd_client/mpd_client_jukebox.c:443: Jukebox: iterating through source, start: 0
DEBUG    mympdapi  /tmp/myMPD-8.1.5/src/mpd_client/mpd_client_jukebox.c:533: Jukebox iterated through 31 songs, skipped 494
WARN     mympdapi  /tmp/myMPD-8.1.5/src/mpd_client/mpd_client_jukebox.c:590: Jukebox queue didn't contain 50 entries
WARN     mympdapi  /tmp/myMPD-8.1.5/src/mpd_client/mpd_client_jukebox.c:592: Disabling jukebox unique constraints temporarily
DEBUG    mympdapi  /tmp/myMPD-8.1.5/src/lib/utility.c:25: Push websocket notify to queue: {"jsonrpc":"2.0","method":"notify","params":{"facility":"jukebox","severity":"warn","message":"Playlist to small, disabling jukebox unique constraints temporarily","data":{}}}
```

> If I input the jukebox parameters in the interface manually, I get normal randomness. The script suppress that randomness.

The ui uses exactly the same api as the your script. You can see the exact api calls with developer console of your browser.

**Comment by jcorporation** [comment, marked as answer]
Concerning the not stopping issue:
- the api call is correctly processes and the stop command is sent to mpd - mpd stops playing
- after the script, the autoplay feature triggers (you have enabled it) because the queue has changed.

You should disable autoplay.

**Comment by thaihugo** [comment]
This is much clearer now. I didn't realize the autoplay was on by default, and I did reinstall Mympd from scratch lately.
As for the playlist, I will augment the number of albums. The fact it was displaying songs as uniqueness factor fooled me. 

Thanks a lot for your great help.
